validator/server_arb/machine.go

```diff
@@ -71,11 +71,14 @@ func (m *ArbitratorMachine) Destroy() {
 		// We no longer need a finalizer
 		runtime.SetFinalizer(m, nil)
 	}
-	m.contextId = nil
-}
 
-func freeContextId(context *int64) {
-	preimageResolvers.Delete(*context)
+	if m.contextId != nil {
+		resolverWithRefCounter, ok := preimageResolvers.Load(*m.contextId)
+		if ok && (resolverWithRefCounter.refCounter.Add(-1) == 0) {
```

Comment:
One other thing, let's log an error if !ok or if the ref count goes below zero. Also, I don't think you need the parens.